go/beacon/api/vrf.go

```diff
@@ -0,0 +1,111 @@
+package api
+
+import (
+	"context"
+
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/signature"
+	"github.com/oasisprotocol/oasis-core/go/common/pubsub"
+	"github.com/oasisprotocol/oasis-core/go/consensus/api/transaction"
+)
+
+// MethodVRFProve is the method name for a VRF proof.
+var MethodVRFProve = transaction.NewMethodName(ModuleName, "VRFProve", VRFProve{})
+
+// VRFParameters are the beacon parameters for the VRF backend.
+type VRFParameters struct {
+	// BaseEpoch is the first epoch from which VRF-based timekeeping
+	// will be active.  VRF proof submissions will be accepted from
+	// `BaseEpoch`, VRF based elections will first occur at `BaseEpoch+1`,
+	// and runtimes will be enabled from `BaseEpoch+2`.
+	BaseEpoch EpochTime `json:"base_epoch,omitempty"`
+
+	// AlphaHighQualityThreshold is the minimum number of proofs (Pi)
+	// that must be received for the next input (Alpha) to be considered
+	// high quality.  If the VRF input is not high quality, runtimes will
+	// be disabled for the next epoch.
+	AlphaHighQualityThreshold uint64 `json:"alpha_hq_threshold,omitempty"`
+
+	// Interval is the epoch interval (in blocks).
+	Interval int64 `json:"interval,omitempty"`
+
+	// ProofSubmissionDelay is the wait peroid in blocks after an epoch
+	// transition that nodes MUST wait before attempting to submit a
+	// VRF proof for the next epoch's elections.
+	ProofSubmissionDelay int64 `json:proof_delay,omitempty`
+}
+
+// VRFState is the VRF backend state.
+type VRFState struct {
+	// Epoch is the epoch for which this alpha is valid.
+	Epoch EpochTime `json:"epoch"`
+
+	// Alpha is the active VRF alpha_string input.
+	Alpha []byte `json:"alpha"`
+
+	// Pi is the accumulated pi_string (VRF proof) outputs.
+	Pi map[signature.PublicKey]*signature.Proof `json:"pi,omitempty`
+
+	// AlphaIsHighQuality is true iff the alpha was generated from
+	// high quality input such that elections will be possible.
+	AlphaIsHighQuality bool `json:"alpha_hq"`
+
+	// SubmitAfter is the block height after which nodes may submit
+	// VRF proofs for the current epoch.
+	SubmitAfter int64 `json:"submit_after"`
+}
+
+// VRFProve is a VRF proof transaction payload.
+type VRFProve struct {
+	Epoch EpochTime `json:"epoch"`
+
+	Pi []byte `json:"pi"`
+}
+
+// MethodMetadata implements transaction.MethodMetadataProvider.
+func (txBody *VRFProve) MethodMetadata() transaction.MethodMetadata {
+	return transaction.MethodMetadata{
+		// XXX: This is critical but any node can submit one, is
+		// this safe to use?
+		Priority: transaction.MethodPriorityCritical,
```

Comment:
If anyone can submit these then this would be unsafe. Critical priority should only be used for things that only proposers can submit.